harp/clock.py

```diff
@@ -0,0 +1,220 @@
+
+    if len(start_times) != len(harp_times):
+        raise ValueError(
+            "The number of start times must equal the number of Harp times"
+        )
+
+    N = len(start_times)
+
+    slopes = np.zeros((N + 1,))
+    offsets = np.zeros((N + 1,))
+
+    # compute overall slope and offset
+    A = np.vstack([start_times, np.ones(len(start_times))]).T
+    slopes[0], offsets[0] = np.linalg.lstsq(A, harp_times, rcond=None)[0]
+
+    # compute slope and offset for each segment
+    for i in range(N):
+        x = start_times[i : i + 2]
+        y = harp_times[i : i + 2]
+        A = np.vstack([x, np.ones(len(x))]).T
+        slopes[i + 1], offsets[i + 1] = np.linalg.lstsq(A, y, rcond=None)[0]
+
+    # find the nearest anchor point for each timestamp to align
+    nearest = np.searchsorted(start_times, timestamps_to_align, side="left")
+
+    nearest[nearest == N] = 0
+
+    # interpolate between the anchor points
+    aligned_times = timestamps_to_align * slopes[nearest] + offsets[nearest]
+
+    return aligned_times
+
+
+def decode_harp_clock(
+    timestamps_or_sample_numbers, states, sample_rate=1, baud_rate=1000.0
+):
+    """
+    Decodes Harp clock times (in seconds) from a sequence of sample
+    numbers and states.
+
+    The Harp Behavior board can be configured to output a digital
+    signal that encodes the current Harp time as a 32-bit integer,
+    which is emitted once per second.
+
+    The format of the signal is as follows:
+
+      Default value: high
+      Start bit: low -- indicates the transition to the next second
+      8 bits: byte 0, with least significant bit first
+      2 bits: high / low transition
+      8 bits: byte 1, with least significant bit first
+      2 bits: high / low transition
+      8 bits: byte 2, with least significant bit first
+      2 bits: high / low transition
+      8 bits: byte 3, with least significant bit first
+      Final bit: reset to high
+
+    Although the baud rate of the internal Harp clock is
+    100 kHz, the Behavior board outputs the clock signal
+    at a lower baud rate (typically 1 kHz), so it can be
+    acquired by data acquisition systems with sample rates
+    as low as 5 kHz.
+
+    Parameters
+    ----------
+    timestamps_or_sample_numbers : np.array
+        Float times in seconds or integer sample numbers
+        for each clock line transition
+    states : np.array
+        States (1 or 0) for each clock line transition
+    sample_rate : float
+        The sample rate at which the clock signal was acquired
+        When inputting timestamps in seconds, use a sample rate of 1
+    baud_rate : float
+        The baud rate of the clock signal
+
+    Returns
+    -------
+    starts : np.array
+        Sample numbers or timestamps at which each second begins
+    harp_times : np.array
+        Harp clock times in seconds
+    """
+
+    min_delta = sample_rate / 2  # 0.5 seconds
+
+    barcode_edges = get_barcode_edges(timestamps_or_sample_numbers, min_delta)
```

Comment:
Since `min_delta` is half of the period in s, I would suggest forcing the use of `timestamps` instead of `timestamps_or_sample_numbers`. What do you think?

Reply:
Yeah, I think just enforcing the use of timestamps in seconds will create less opportunities for user errors.

Reply:
This is fixed in the latest set of commits